lib/Doctrine/ODM/MongoDB/Iterator/BaseIterator.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ODM\MongoDB\Iterator;
+
+use Generator;
+use Traversable;
+
+/**
+ * Iterator for wrapping a Traversable and caching its results.
+ *
+ * By caching results, this iterators allows a Traversable to be counted and
+ * rewound multiple times, even if the wrapped object does not natively support
+ * those operations (e.g. MongoDB\Driver\Cursor).
+ *
+ * @internal
+ */
+final class BaseIterator implements Iterator
```

Comment:
`BaseIterator` makes it sounds like this is designed to be extended, which it isn't due to it being `final` and `@internal`. The SPL brings a `NoRewindIterator`, which would be a good name except that it behaves differently (it ignores calls to `rewind` instead of throwing an exception. What do you think of calling this `UnrewindableIterator`?

```suggestion
final class UnrewindableIterator implements Iterator
```

Reply:
Applied alongside other comments related to `UnrewindableIterator` (previously `BaseIterator`) in 907e08d5bee727860eac12ef51dc6392a654c613.